### PR TITLE
Build System Redesign

### PR DESCRIPTION
**Pull request by Kfeavel**
Still going to stick with Make since it provides the greatest level of control, but we should figure out how we want to better integrate source files as our project starts to include libraries, other programs (when we get ELF loading), etc.

### CODE REVIEW

File: kernel/sys/early_panic.cpp [hunk 1]
```diff
@@ -12,9 +12,12 @@
 #include <arch/arch.hpp>
 #include <dev/tty/tty.hpp>
 
+// Provide a function prototype to make the compiler warnings happy. Don't
+// want to make it public though, so we won't put it in the panic header.
+extern "C" void early_panic(const char *str);
+
 extern "C" void
 __attribute__ ((section(".early_text")))
-__attribute__ ((optnone))
```
Comment (micahswitzer) [inline kernel/sys/early_panic.cpp:17]: Why was this removed? Won't that break this functionality?

File: Makefile [hunk 2]
```diff
@@ -41,26 +42,37 @@ QEMU = $(shell command -v qemu-system-$(QEMU_ARCH))
...
+	-Wshadow             \
+	-Wcast-align         \
+	-Wno-long-long       \
+	-Wpointer-arith      \
+	-Wconversion         \
+	-Wwrite-strings      \
+	-Wredundant-decls    \
+	-Wmissing-declarations
+# TODO: Include -Werror once we
+# Fix all of the warnings
```
Comment (micahswitzer) [inline Makefile]: Have we not fixed all the warnings?

File: Makefile [hunk 3]
```diff
@@ -102,10 +129,15 @@ OBJ_DIRS = $(subst kernel, obj, $(shell find kernel -type d))
 mkdir_obj_dirs:
 	mkdir -p $(OBJ_DIRS)
 
+# C source -> object
+obj/%.o: kernel/%.c $(HEADERS)
+	$(MAKE) mkdir_obj_dirs
+	$(CXX) $(CFLAGS) $(CPPFLAGS) -c -o $@ $<
```
Comment (micahswitzer) [inline Makefile]: Shouldn't it be `$(CC)` for C code?

File: kernel/lib/printf.cpp [hunk 4]
```diff
@@ -288,6 +294,10 @@ OK, I found my mistake. The math here is _always_ unsigned */
             default:
                 break;
             }
+            // Tell the compiler that this isn't an issue
+            // (at least I'm assuming it isn't)
+            [[fallthrough]];
```
Comment (micahswitzer) [inline kernel/lib/printf.cpp:299]: Maybe explain to the person looking at the code why this is the case as well

File: kernel/lib/semaphore.cpp [hunk 5]
```diff
@@ -131,6 +132,8 @@ int px_sem_post(px_sem_t *sem) {
 }
 
 uint32_t px_sem_getval(px_sem_t *sem, uint32_t *val) {
+    // FIXME: I don't think val should be unused?
+    (void)val;
```
Comment (micahswitzer) [inline kernel/lib/semaphore.cpp:136]: Probably not. Can you make an issue for this?

File: kernel/mem/paging.cpp [hunk 6]
```diff
@@ -88,7 +94,7 @@ static void px_paging_init_dir() {
         px_map_kernel_page_table(i, &page_tables[i]);
         // clear out the page tables
         for (int j = 0; j < PAGE_ENTRIES; j++) {
-            page_tables[i].pages[j] = (px_page_table_entry_t){ 0 };
+            page_tables[i].pages[j] = (px_page_table_entry_t){ /* ZERO */ };
```
Comment (micahswitzer) [inline kernel/mem/paging.cpp:97]: I prefer the actual zero, but maybe you did this because of warnings?

File: kernel/sys/early_panic.cpp (hunk 1 above)
Comment (Kfeavel) [reply kernel/sys/early_panic.cpp:17]: GCC complained about it. Apparently it's a clang only thing.

File: Makefile (hunk 2 above)
Comment (Kfeavel) [reply Makefile]: Nope, because we added more with these new flags. Most of the ones left are due to implicit casting.

File: Makefile (hunk 3 above)
Comment (Kfeavel) [reply Makefile]: Fixed in the next commit.

File: kernel/lib/printf.cpp (hunk 4 above)
Comment (Kfeavel) [reply kernel/lib/printf.cpp:299]: I'm not certain but I think it's because we want the code below to execute if it falls through so that we continue to print. If you have a better guess / comment just leave it here and I'll add it.

File: kernel/lib/semaphore.cpp (hunk 5 above)
Comment (Kfeavel) [reply kernel/lib/semaphore.cpp:136]: #151 is now open

File: kernel/mem/paging.cpp (hunk 6 above)
Comment (Kfeavel) [reply kernel/mem/paging.cpp:97]: Yes, I did this because of the warnings. In C this would set everything to 0, but in C++ this only sets the first struct member to 0. (The rest are technically set to 0, but the compiler complains since the correct way is to leave the `{ }` blank.

File: kernel/sys/early_panic.cpp (hunk 1 above)
Comment (micahswitzer) [reply kernel/sys/early_panic.cpp:17]: Okay so we'll need to add the appropriate attribute for GCC. I believe it is something like `__attribute__((-O0))`.

File: kernel/mem/paging.cpp (hunk 6 above)
Comment (micahswitzer) [reply kernel/mem/paging.cpp:97]: Ooo. I did not know that. I kinda don't like it, but I guess it is how it is. I'm also not a fan of the comment, but that's whatever.

Comment (Kfeavel) [reply kernel/mem/paging.cpp:97]: I can remove the comments, but I figured it would be better to have something explaining that it's all zero'd out.

File: kernel/lib/printf.cpp (hunk 4 above)
Comment (micahswitzer) [reply kernel/lib/printf.cpp:299]: Ah okay yeah. It's kinda a horribly structured function in the first place.

File: kernel/sys/early_panic.cpp (hunk 1 above)
Comment (Kfeavel) [reply kernel/sys/early_panic.cpp:17]: `__attribute__((-O0))` doesn't work. I'll see if I can't find what it is.

File: Makefile (hunk 2 above)
Comment (micahswitzer) [reply Makefile]: Is there any reason they can't be addressed now?

File: kernel/sys/early_panic.cpp (hunk 1 above)
Comment (Kfeavel) [reply kernel/sys/early_panic.cpp:17]: Is `__attribute__((optimize("O0")))` what you're looking for?

File: Makefile (hunk 2 above)
Comment (Kfeavel) [reply Makefile]: I'm working on it. I've already fixed the 90% of other issues but type casting gets tricky, especially when we're dealing with structs which only use 20 bits of an int and stuff. (Yes, it does complain about that).

File: kernel/lib/printf.cpp (hunk 4 above)
Comment (Kfeavel) [reply kernel/lib/printf.cpp:299]: Yeah, it is, but I don't feel like optimizing it right now.

File: kernel/sys/early_panic.cpp (hunk 1 above)
Comment (micahswitzer) [reply kernel/sys/early_panic.cpp:17]: Probs

File: kernel/lib/printf.cpp (hunk 4 above)
Comment (micahswitzer) [reply kernel/lib/printf.cpp:299]: All good

File: kernel/mem/paging.cpp (hunk 6 above)
Comment (micahswitzer) [reply kernel/mem/paging.cpp:97]: Ok

File: kernel/sys/early_panic.cpp (hunk 1 above)
Comment (Kfeavel) [reply kernel/sys/early_panic.cpp:17]: Yeah, it is. Found a stackoverflow post about it.

File: Makefile (hunk 2 above)
Comment (Kfeavel) [reply Makefile]: For right now `-Wconversions` has been removed since that was the only flag that was giving us trouble, and I'm not sure if it's something we want to keep. If it is then it can be added back, but we'll have to fix a **ton** of casting warnings

**Review by micahswitzer — APPROVED** [review]
